lib/internal/fetch/headers.js

```diff
@@ -0,0 +1,212 @@
+'use strict';
+
+const {
+  codes: { ERR_INVALID_ARG_VALUE },
+} = require('internal/errors');
+
+const {
+  Array,
+  ArrayIsArray,
+  MathFloor,
+  ObjectEntries,
+  StringPrototypeReplace,
+  StringPrototypeToLocaleLowerCase,
+  Symbol,
+  SymbolIterator,
+} = primordials;
+
+const { validateObject } = require('internal/validators');
+const { isBoxedPrimitive } = require('internal/util/types');
+
+const { validateHeaderName, validateHeaderValue } = require('_http_outgoing');
+
+const { Buffer } = require('buffer');
+
+const kHeadersList = Symbol('headers list');
+
+/**
+ * This algorithm is based off of
+ * https://www.tbray.org/ongoing/When/200x/2003/03/22/Binary
+ * It only operates on the even indexes of the array (the header names) by only
+ * iterating at most half the length of the input array. The search also
+ * assumes all entries are strings and uses String.prototype.localeCompare for
+ * comparison.
+ */
+function binarySearch(arr, val) {
+  let low = 0;
+  let high = MathFloor(arr.length / 2);
+
+  while (high > low) {
+    const mid = (high + low) >>> 1;
+
+    if (val.localeCompare(arr[mid * 2]) > 0) {
+      low = mid + 1;
+    } else {
+      high = mid;
+    }
+  }
+
+  return low * 2;
+}
+
+function normalizeAndValidateHeaderName(name) {
+  const normalizedHeaderName = StringPrototypeToLocaleLowerCase(name);
+  validateHeaderName(normalizedHeaderName);
+  return normalizedHeaderName;
+}
+
+function normalizeAndValidateHeaderValue(name, value) {
+  // https://fetch.spec.whatwg.org/#concept-header-value-normalize
+  const normalizedHeaderValue = StringPrototypeReplace(
+    value,
+    /^[\n\t\r\x20]+|[\n\t\r\x20]+$/g,
+    ''
+  );
```

Comment:
```suggestion
  const normalizedHeaderValue = RegExpPrototypeSymbolReplace(
    /^[\n\t\r\x20]+|[\n\t\r\x20]+$/g,
    value,
    ''
  );
```

Reply:
Any reason for this recommendation? Is RegExp replace faster than String replace?

Reply:
`%String.prototype.replace%` simply calls `regex[Symbol.replace](this, replacer)`, so you can skip a step by using the RegExp method directly.